File: dart/packages/aoc2021/lib/src/day07.dart
// https://adventofcode.com/2021/day/7

import 'package:aoc/aoc.dart';
import 'package:aoc/util/range.dart';
import 'package:collection/collection.dart';

main() => Day07().solve();

class Day07 extends AdventDay {
  Day07() : super(2021, 7);

  @override
  dynamic part1(String input) {
    final List<int> crabs = parseCrabs(input);
    return rangeinc(crabs.min, crabs.max).map((t) => fuelUsed(crabs, t)).min;
  }

  @override
  dynamic part2(String input) {
    final crabs = parseCrabs(input);
    return rangeinc(crabs.min, crabs.max)
        .map((t) => acceleratingFuelUsed(crabs, t))
        .min;
  }

  static List<int> parseCrabs(String input) =>
      input.split(',').map(int.parse).toList();

  static int fuelUsed(Iterable<int> crabs, int target) =>
      crabs.map((c) => (target - c).abs()).sum;

  static int acceleratingFuelUsed(Iterable<int> crabs, int target) =>
      crabs.map((c) {
        final dist = (target - c).abs();
        return dist * (dist + 1) ~/ 2;
      }).sum;
}

File: clojure/src/aoc2022/day07.clj
;; https://adventofcode.com/2022/day/7
(ns aoc2022.day07
  (:require
   [aoc.day :as d]
   [aoc.util.math :refer [sum]]
   [aoc.util.memoize :refer [letfn-mem]]
   [aoc.util.string :as s]
   [clojure.string :as str]))

(defn input [] (d/day-input 2022 7))

(defn- path-for [path dir-name]
  (conj (or path []) dir-name))

(defn- cd [cwd dir-name]
  (case dir-name
    "/" ["/"]
    ".." (pop cwd)
    (path-for cwd dir-name)))

(defn- process-line [fs line]
  (let [cwd (:cwd fs)
        stripped (str/replace line #"^\$ " "")
        [cmd arg] (str/split stripped #" ")]
    (case cmd
      "cd"  (assoc fs :cwd (cd cwd arg))
      "ls"  fs
      "dir" (update-in fs [cwd :dirs] conj arg)
      (update-in fs [cwd :files-size] (fnil + 0) (s/int cmd)))))

(defn- parse-output [input]
  (dissoc (reduce process-line {} (s/lines input)) :cwd))

(defn- dir-sizes [fs]
  (letfn-mem
   [(size-of [path]
             (let [dir (fs path)]
               (+ (:files-size dir 0)
                  (sum (map #(size-of (path-for path %))
                            (:dirs dir))))))]
   (map size-of (keys fs))))

(defn part1 [input]
  (->> (parse-output input)
       dir-sizes
       (filter #(<= % 100000))
       sum))

(defn part2 [input]
  (let [sizes (dir-sizes (parse-output input))
        root-size (apply max sizes)
        needed-space (- root-size (- 70000000 30000000))]
    (->> sizes
         (filter #(>= % needed-space))
         (apply min))))
